### PR TITLE
Make Megatron-FSDP torch.compile compatible

### PR DESCRIPTION
**Pull request by shjwudp**
# What does this PR do ?

This PR makes Megatron-FSDP compatible with `torch.compile` by disabling compilation for its internal FSDP hooks using `@torch.compiler.disable`. These hooks rely on eager-mode behavior and can confuse the compiler’s tracing and graph construction, leading to graph breaks or errors when compiling models wrapped with Megatron-FSDP. By explicitly opting these hook entry points out of compilation, the main model computation remains compilable while Megatron-FSDP continues to manage sharding and communication in eager mode.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py [hunk 1]
```diff
@@ -235,6 +235,7 @@ def teardown_class(cls):
     )
     @pytest.mark.parametrize("preserve_fp32_weights", [True, False])
     @pytest.mark.parametrize("init_model_with_meta_device", [True, False])
+    @pytest.mark.parametrize("torch_compile", [True, False])
```
Comment (cspades) [inline tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py]: @shjwudp Nit: Is it possible to have a separate unit test for `torch.compile` or do we need to test it on the full matrix of Megatron-FSDP features? (At some point, this test will OOM from memory orphans. All of the existing features capture important edge cases so they should be tested together, I think compilation is a very orthogonal function to the above arguments.)

**Comment by cspades** [comment]
Clicked the "update branch" button since main branch CI was broken. CI should pass now.

**Comment by cspades** [comment]
/ok to test [eabf4fe](https://github.com/NVIDIA/Megatron-LM/pull/2425/commits/eabf4fe6b826c39c688dd599fb0481213cbf0703)

**Comment by cspades** [comment]
/ok to test [b25628d](https://github.com/NVIDIA/Megatron-LM/pull/2425/commits/b25628de03681015df345e0881de9e2cc1c98c4a)

**Review by jaredcasper — APPROVED** [review]

**Comment by cspades** [comment]
One more try to get CI to pass...

**Comment by cspades** [comment]
/ok to test [59f6f83](https://github.com/NVIDIA/Megatron-LM/pull/2425/commits/59f6f83f96f4a34c348d8857891e07dd447eb895)

**Comment by cspades** [comment]
What is going on with main branch 👀 
```
2026-01-15T16:49:32.7940913Z 1-task-1-0/0 [default0]:FAILED tests/unit_tests/models/test_mamba_moe_model.py::TestMambaMoEModel::test_constructor
```
No way this is related to our changes...

**Comment by cspades** [comment]
Updated branch, should be fixed by: https://github.com/NVIDIA/Megatron-LM/pull/2970

**Comment by cspades** [comment]
/ok to test [14080c1](https://github.com/NVIDIA/Megatron-LM/pull/2425/commits/14080c197ed4f52e8b297aa272caf157202a56d3)

**Comment by cspades** [comment]
/ok to test [cad4126](https://github.com/NVIDIA/Megatron-LM/pull/2425/commits/cad41265963180b2991fb822f640487a2725d350)
